gen/map.go

```diff
@@ -281,3 +281,147 @@ func (m *mapGenerator) equalsUnhashable(g Generator, spec *compile.MapSpec) (str
 
 	return name, wrapGenerateError(spec.ThriftName(), err)
 }
+
+func (m *mapGenerator) zapMarshaler(
+	g Generator,
+	root *compile.MapSpec,
+	fieldValue string,
+) (string, error) {
+	name := zapperName(g, root)
+	switch compile.RootTypeSpec(root.KeySpec).(type) {
+	case *compile.StringSpec:
+		return m.zapStringKeyMarshaler(g, name, root, fieldValue)
+	default:
+		return m.zapNonstringKeyMarshaler(g, name, root, fieldValue)
+	}
+}
+
+func (m *mapGenerator) zapStringKeyMarshaler(
+	g Generator,
+	name string,
+	root *compile.MapSpec,
+	fieldValue string,
+) (string, error) {
+	if err := g.EnsureDeclared(
+		`
+			<$zapcore := import "go.uber.org/zap/zapcore">
+
+			type <.Name> <typeReference .Type>
+			<$m := newVar "m">
+			<$k := newVar "k">
+			<$v := newVar "v">
+			<$enc := newVar "enc">
+			// MarshalLogObject implements zapcore.ObjectMarshaler, enabling
+			// fast logging of <.Name>.
+			func (<$m> <.Name>) MarshalLogObject(<$enc> <$zapcore>.ObjectEncoder) error {
+				for <$k>, <$v> := range <$m> {
+					<zapEncodeBegin .Type.ValueSpec ->
+						<$enc>.Add<zapEncoder .Type.ValueSpec>((string)(<$k>), <zapMarshaler .Type.ValueSpec $v>)
```

Comment:
If `$k` is already a string, do we need the `(string)` here?

Reply:
It may be a typedef of a string or a typedef of a typedef of a string.